src/passes/TypeSSA.cpp

```diff
@@ -0,0 +1,204 @@
+// allowing us to infer what is printed.
+//
+// Note that the metaphor of SSA is not perfect here as we do not handle merges,
+// that is, when there is a value that can read from more than one struct.new
+// then we do nothing atm. We could create a phi there, but in general that
+// would require multiple inheritance. TODO think more on that
+//
+
+#include "ir/find_all.h"
+#include "ir/module-utils.h"
+#include "ir/names.h"
+#include "ir/possible-constant.h"
+#include "pass.h"
+#include "wasm-builder.h"
+#include "wasm.h"
+
+namespace wasm {
+
+namespace {
+
+struct News {
+  std::vector<StructNew*> structNews;
+  // TODO: arrayNews of all kinds
+};
+
+struct NewFinder : public PostWalker<NewFinder> {
+  News news;
+
+  void visitStructNew(StructNew* curr) { news.structNews.push_back(curr); }
+
+  // TODO arrays
+};
+
+struct TypeSSA : public Pass {
+  // Only modifies struct/array.new types.
+  bool requiresNonNullableLocalFixups() override { return false; }
+
+  Module* module;
+
+  void run(Module* module_) override {
+    module = module_;
+
+    if (!module->features.hasGC()) {
+      return;
+    }
+    if (getTypeSystem() != TypeSystem::Nominal &&
+        getTypeSystem() != TypeSystem::Isorecursive) {
+      Fatal() << "TypeSSA requires nominal/isorecursive typing";
+    }
+
+    // First, find all the struct/array.news.
+
+    ModuleUtils::ParallelFunctionAnalysis<News> analysis(
+      *module, [&](Function* func, News& news) {
+        if (func->imported()) {
+          return;
+        }
+
+        NewFinder finder;
+        finder.walk(func->body);
+        news = std::move(finder.news);
+      });
+
+    // Also find news in the module scope.
+    NewFinder moduleFinder;
+    moduleFinder.walkModuleCode(module);
+
+    // Process all the news. Note that we must do so in a deterministic order.
+    ModuleUtils::iterDefinedFunctions(
+      *module, [&](Function* func) { processNews(analysis.map[func]); });
+    processNews(moduleFinder.news);
+  }
+
+  // As we generate new names, use a consistent index.
+  Index nameCounter = 0;
+
+  void processNews(const News& news) {
+    // We'll generate nice names as we go, if we can.
+    std::unordered_set<Name> existingTypeNames;
+    auto& typeNames = module->typeNames;
+    for (auto& [type, info] : typeNames) {
+      existingTypeNames.insert(info.name);
+    }
+
+    for (auto* curr : news.structNews) {
+      if (isInteresting(curr)) {
+        // This is interesting; create a new type and use it. The new type is
+        // identical to the existing one, and a subtype of it, so the only
+        // difference between them is nominal - which is enough for optimization
+        // passes to learn different things about the two.
+        auto oldType = curr->type.getHeapType();
+        TypeBuilder builder(1);
+        builder.setHeapType(0, oldType.getStruct());
+        builder.setSubType(0, oldType);
```

Comment:
Oh, yikes, good catch. That's pretty bad actually - sounds like we'd need to accumulate all the relevant news and handle them all at once. Perhaps the pass could only work in nominal mode for now, though - do we have users of non-nominal mode, do you know?

Reply:
I assume the ocaml folks are using the isorecursive system, and in general we currently have the nice property that you get pretty much the same optimizations in isorecursive and nominal modes. I think it would be worth supporting isorecursive types now.

Reply:
> The problem is that we don't want to create a new rec group that happens to match an existing one...

It's possible that the new rec group matches an existing rec group, but in that case none of the types in the existing rec group could possible have been allocated. (if they had been allocated, they wouldn't be in the group of new subtypes of allocated types!) It's still possible that there exists some cast to the existing type, though, which could cause problems. I seriously doubt this would happen in practice, so I think it would be fine to detect whether one of the "new" types is already used by the module and bail out of the optimization if that's the case.

> a single use of a type from that huge new rec group would keep the entire thing alive, wasting space...

This could be a problem even without hacks where we make huge rec groups just to make them "nominal." In a closed world, we could be optimizing unused types out of rec groups as long as we can prove that removing the types won't unify groups that would otherwise be distinct.

Reply:
> It's possible that the new rec group matches an existing rec group, but in that case none of the types in the existing rec group could possible have been allocated. (if they had been allocated, they wouldn't be in the group of new subtypes of allocated types!) 

I don't follow that, sorry. (What does "allocated" mean? `struct.new`? But even if so, I don't follow.)

To elaborate what I think is the problem, say we have existing an existing rec group:
```wat
(rec
  (type $existing (struct_subtype (..fields of $A, no changes..) $A))
)
```
That rec group just happens to exist right now. And lets say we have a `struct.new $A` and we want to create a new type for that new, so we create
```wat
(rec
  (type $intended.to.be.new (struct_subtype (..fields of $A, no changes..) $A))
)
```
That is structurally identical to the first rec group, so `$existing` and `$intended.to.be.new` would be equal, I think?

And that is regardless of how `$existing` is used in the program. We'd be switching that `struct.new $A` to `struct.new $existing`, and old uses of that type could include things that prevent optimizations that we could do if we used a new type here (`$existing` could be written something generic, while the `struct.new $A` might have only sets of something specific/refined).

Reply:
Yes, that's another problem with creating one new type at a time. They will be duplicates of each other, but also potentially duplicates of other types that already exist. I suppose if you create new subtypes for types that are only ever allocated once you can end up with a similar problem.

To minimize the chances of getting a conflicting with existing types, all the new types this pass will add for the entire module could be part of a single large rec group. I think that's what I had in mind when I wrote "none of the types in the existing rec group could possibl[y] have been allocated. (if they had been allocated, they wouldn't be in the group of new subtypes of allocated types!)" Yes, by allocated I mean allocated with a struct.new instruction.

Reply:
I believe the problem I described can still happen while generating a single big rec group for all types at once, since we don't create a new type for all `struct.new`s. I think I see now that you were assuming that. But the pass actually finds news that can't be optimized ("uninteresting" ones) and does nothing on them. So I might find I want to create 10 new fresh subtypes of `$A, $B` and maybe, by chance, those types already have 10 existing subtypes - none of which are interesting - and those existing subtypes are in a rec group with the exact same shape as the one the pass creates :scream: 

IIUC the only real solution is to actually manually verify no existing rec group can possibly be structurally identical. Like by make a rec group larger than all others, or with a new weird element with more fields than all others. Which are all quite horrible...

Reply:
Ah, I see, thanks. I still think that bailing out of the optimization if any of the new types are already used by the module would be an acceptable solution since it should be sufficiently rare (I hope) in practice.

Reply:
Ok, code rewritten to create a single big rec group. I decided to error if we fail to create new/private/fresh types, for now - seems safer, and hopefully very rare as you said.